File: AlerteTCL/Services/TCLAPIService.swift
import Foundation

actor TCLAPIService {
    static let shared = TCLAPIService()
    
    // Proxy Cloudflare Worker — les credentials Grand Lyon sont stockés
    // côté Cloudflare en secrets chiffrés, jamais dans le binaire de l'app.
    private let alertsEndpoint = NetworkConfiguration.proxyBaseURL + "/alerts"
    
    private init() {}
    
    func fetchAlerts() async throws -> [TCLAlert] {
        guard let url = URL(string: alertsEndpoint) else {
            AppLogger.debug("❌ TCLAPIService: URL invalide")
            throw ServiceError.invalidURL
        }
        
        AppLogger.debug("🌐 TCLAPIService: Chargement des alertes depuis le proxy")
        
        var request = NetworkConfiguration.request(url: url, timeout: NetworkConfiguration.fastTimeout)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("AlerteTCL/1.0", forHTTPHeaderField: "User-Agent")
        AppLogger.debug("🚀 Alertes: Début requête (timeout: \(NetworkConfiguration.fastTimeout)s)")
        
        // Pas d'auth côté app — le proxy Cloudflare gère les credentials
        
        let (data, response) = try await NetworkConfiguration.session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            AppLogger.debug("❌ TCLAPIService: Réponse invalide")
            throw ServiceError.invalidResponse
        }
        
        AppLogger.debug("📡 TCLAPIService: Status code: \(httpResponse.statusCode)")
        
        guard httpResponse.statusCode == 200 else {
            if httpResponse.statusCode == 401 {
                AppLogger.debug("❌ TCLAPIService: Authentification échouée")
                throw ServiceError.unauthorized
            }
            AppLogger.debug("❌ TCLAPIService: Erreur HTTP \(httpResponse.statusCode)")
            throw ServiceError.httpError(httpResponse.statusCode)
        }
        
        do {
            let decoder = JSONDecoder()
            let apiResponse = try decoder.decode(APIResponse.self, from: data)
            AppLogger.debug("✅ TCLAPIService: \(apiResponse.values.count) alertes chargées")
            
            // Filtrer les alertes actives
            let activeAlerts = apiResponse.values.filter { $0.isActive }
            AppLogger.debug("📊 TCLAPIService: \(activeAlerts.count) alertes actives")
            
            return activeAlerts
        } catch {
            AppLogger.debug("❌ TCLAPIService: Erreur de décodage: \(error)")
            throw ServiceError.decodingError(error)
        }
    }

    func fetchBusLineNames() async throws -> [TransportLine] {
        guard let url = URL(string: NetworkConfiguration.proxyBaseURL + "/bus-termini") else {
            throw ServiceError.invalidURL
        }
        var request = NetworkConfiguration.request(url: url, timeout: NetworkConfiguration.fastTimeout)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("AlerteTCL/1.0", forHTTPHeaderField: "User-Agent")

        let (data, response) = try await NetworkConfiguration.session.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw ServiceError.invalidResponse
        }

        let decoded = try JSONDecoder().decode(BusTerminiResponse.self, from: data)
        let names = Set(decoded.features.compactMap { $0.properties.ligne })
        return names.sorted().map { name in
            let mode: TransportMode = name.hasPrefix("C") ? .busC : .bus
            return TransportLine(ligneCom: name, ligneCli: name, mode: mode)
        }
    }
}

// MARK: - Bus-termini response models (private)

private struct BusTerminiResponse: Decodable {
    let features: [BusTerminiFeature]
}

private struct BusTerminiFeature: Decodable {
    let properties: BusTerminiProperties
}

private struct BusTerminiProperties: Decodable {
    let ligne: String?
}

File: shared/src/commonMain/kotlin/com/alertetcl/shared/services/TclApiService.kt
package com.alertetcl.shared.services

import com.alertetcl.shared.models.TCLAlert
import com.alertetcl.shared.models.TransportMode
import com.alertetcl.shared.network.ApiError
import com.alertetcl.shared.network.HttpClientProvider
import com.alertetcl.shared.network.NetworkConfiguration
import com.alertetcl.shared.network.dto.AlertsApiResponse
import com.alertetcl.shared.util.AppLogger
import com.alertetcl.shared.util.parseIsoEpoch
import io.ktor.client.call.body
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.client.statement.HttpResponse
import io.ktor.http.HttpStatusCode
import kotlinx.datetime.Clock

/**
 * Service de chargement des alertes trafic TCL via le proxy Cloudflare.
 */
class TclApiService {

    private val endpoint = NetworkConfiguration.PROXY_BASE_URL + "/alerts"
    private val client = HttpClientProvider.client

    suspend fun fetchAlerts(): List<TCLAlert> {
        val response: HttpResponse = try {
            client.get(endpoint) {
                timeout { requestTimeoutMillis = NetworkConfiguration.FAST_TIMEOUT_SECONDS * 1000 }
            }
        } catch (e: Throwable) {
            AppLogger.error("TclApiService: erreur réseau", e)
            throw ApiError.NetworkError(e)
        }

        when (response.status) {
            HttpStatusCode.OK -> { /* continue */ }
            HttpStatusCode.Unauthorized -> throw ApiError.Unauthorized
            else -> throw ApiError.HttpError(response.status.value)
        }

        val body: AlertsApiResponse = try {
            response.body()
        } catch (e: Throwable) {
            throw ApiError.DecodingError(e)
        }

        val now = Clock.System.now().epochSeconds
        val alerts = body.values.mapNotNull { dto ->
            val id = "${dto.n ?: 0}-${dto.ligneCom.orEmpty()}-${dto.ligneCli.orEmpty()}"
            TCLAlert(
                id = id,
                type = dto.type ?: "Information",
                cause = dto.cause ?: "",
                debutEpoch = parseIsoEpoch(dto.debut),
                finEpoch = parseIsoEpoch(dto.fin),
                mode = TransportMode.fromString(dto.mode),
                ligneCom = dto.ligneCom ?: "",
                ligneCli = dto.ligneCli ?: "",
                titre = dto.titre ?: "",
                message = dto.message ?: "",
                niveauSeverite = dto.niveauSeverite
            )
        }.filter { it.isActive(now) }

        AppLogger.debug("TclApiService: ${alerts.size} alertes actives")
        return alerts
    }

    companion object { val shared = TclApiService() }
}
